File: tests/uasm/sample/boolean_shortcircuit_sideeffects/BooleanShortcircuitSideeffectsTest.cs
using UdonSharp;
using UnityEngine;

[UdonBehaviourSyncMode(BehaviourSyncMode.None)]
public class BooleanShortcircuitSideeffectsTest : UdonSharpBehaviour
{
    private int counter = 0;

    private bool BumpTrue()
    {
        counter += 1;
        return true;
    }

    private bool BumpFalse()
    {
        counter += 1;
        return false;
    }

    public void Start()
    {
        bool first = BumpFalse() && BumpTrue();
        bool second = BumpTrue() || BumpFalse();

        Debug.Log(counter);
        Debug.Log(first);
        Debug.Log(second);
    }
}

File: tests/uasm/sample/boolean_shortcircuit_sideeffects/BooleanShortcircuitSideeffectsTest.ts
import { UdonBehaviour } from "@ootr/udon-assembly-ts/stubs/UdonDecorators";
import { UdonSharpBehaviour } from "@ootr/udon-assembly-ts/stubs/UdonSharpBehaviour";
import type { UdonInt } from "@ootr/udon-assembly-ts/stubs/UdonTypes";
import { Debug } from "@ootr/udon-assembly-ts/stubs/UnityTypes";

@UdonBehaviour()
export class BooleanShortcircuitSideeffectsTest extends UdonSharpBehaviour {
  private counter: UdonInt = 0 as UdonInt;

  private bumpTrue(): boolean {
    this.counter = (this.counter + (1 as UdonInt)) as UdonInt;
    return true;
  }

  private bumpFalse(): boolean {
    this.counter = (this.counter + (1 as UdonInt)) as UdonInt;
    return false;
  }

  Start(): void {
    const first = this.bumpFalse() && this.bumpTrue();
    const second = this.bumpTrue() || this.bumpFalse();

    Debug.Log(this.counter);
    Debug.Log(first);
    Debug.Log(second);
  }
}
